language/compiler/bytecode_source_map/src/disassembler.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::mapping::SourceMapping;
+use crate::source_map::SourceName;
+use failure::prelude::*;
+use libra_types::identifier::IdentStr;
+use vm::access::ModuleAccess;
+use vm::file_format::{
+    FieldDefinitionIndex, FunctionDefinition, FunctionDefinitionIndex, Kind, SignatureToken,
+    StructDefinition, StructDefinitionIndex, StructFieldInformation, TableIndex, TypeSignature,
+};
+
+/// Holds the various options that we support while disassembling code.
+#[derive(Debug, Default)]
+pub struct DisassemblerOptions {
+    /// Only print public functions
+    pub only_public: bool,
+
+    // TODO: Need to implement this.
+    /// Print the bytecode for the instructions within the function.
+    pub print_code: bool,
+}
+
+impl DisassemblerOptions {
+    pub fn new() -> Self {
+        Self {
+            only_public: false,
+            print_code: false,
+        }
+    }
+
+    pub fn only_public(&mut self) {
+        self.only_public = true;
+    }
+
+    pub fn print_code(&mut self) {
+        self.print_code = true;
+    }
+}
+
+pub struct Disassembler<Location: Clone + Eq + Default> {
+    source_mapper: SourceMapping<Location>,
+    // The various options that we can set for disassembly.
+    options: DisassemblerOptions,
+}
+
+impl<Location: Clone + Eq + Default> Disassembler<Location> {
+    pub fn new(source_mapper: SourceMapping<Location>, options: DisassemblerOptions) -> Self {
+        Self {
+            source_mapper,
+            options,
+        }
+    }
+
+    //***************************************************************************
+    // Helpers
+    //***************************************************************************
+
+    fn get_function_def(
+        &self,
+        function_definition_index: FunctionDefinitionIndex,
+    ) -> Result<&FunctionDefinition> {
+        if function_definition_index.0 as usize >= self.source_mapper.bytecode.function_defs().len()
+        {
+            bail!("Invalid function definition index supplied when marking function")
```

Comment:
this is good for now obviously but let's think about how to report those errors.
The index of where this is happening is usually a good piece of info to give back.
No change here to do now, this is more something for us to think about.
Of course same goes for the other `bail!`

Reply:
Absolutely agreed. Once the bytecode printing gets in the next place is looking at reporting these errors back out as well.